# Q-Learning Intro with Frozen Lake

Code can be found at: https://deeplizard.com/learn/video/QK_PP_2KgGE

## 1. Setup

In [ ]:
import numpy as np
import gym
import random
import time
from IPython.display import clear_output

In [ ]:
env = gym.make("FrozenLake-v0")

### Create Q-Table

In [ ]:
action_space_size = env.action_space.n
state_space_size = env.observation_space.n

q_table = np.zeros((state_space_size, action_space_size))

In [ ]:
# Display Env
env.render()

In [ ]:
# Display Q-Table
# Note: Because there are 16 possible spaces and 4 possible actions per space, the Q-Table is 16x4
print(q_table)

### Set Parameters

In [6]:
# Number of Episodes
num_episodes = 10000
# Maximum amount of moves per episode
max_steps_per_episode = 100

# Learning Rate, Alpha
learning_rate = 0.1
# Discount Rate, Gamma
discount_rate = 0.99

exploration_rate = 1
max_exploration_rate = 1
min_exploration_rate = 0.01
exploration_decay_rate = 0.01

rewards_all_episodes = []

## 2. Q-Learning

In [7]:
# Q-learning algorithm
for episode in range(num_episodes):
    # initialize new episode params
    state = env.reset()
    done = False
    rewards_current_episode = 0
    
    for step in range(max_steps_per_episode): 
        # Exploration-exploitation trade-off
        exploration_rate_threshold = random.uniform(0, 1)
        if exploration_rate_threshold > exploration_rate:
            action = np.argmax(q_table[state,:]) 
        else:
            action = env.action_space.sample()
        # Take new action
        new_state, reward, done, info = env.step(action)
        # Update Q-table
        q_table[state, action] = q_table[state, action] * (1 - learning_rate) + \
            learning_rate * (reward + discount_rate * np.max(q_table[new_state, :]))
        # Set new state
        state = new_state
        # Add new reward
        rewards_current_episode += reward
        
        if done == True: 
            break
        
    # Exploration rate decay
    exploration_rate = min_exploration_rate + \
        (max_exploration_rate - min_exploration_rate) * np.exp(-exploration_decay_rate*episode)
    # Add current episode reward to total rewards list
    rewards_all_episodes.append(rewards_current_episode)
    env.reset()

## 3. Results

In [8]:
# Calculate and print the average reward per thousand episodes
rewards_per_thousand_episodes = np.split(np.array(rewards_all_episodes),num_episodes/1000)
count = 1000

print("Average reward per thousand episodes: \n")
for r in rewards_per_thousand_episodes:
    print(count, ": ", str(sum(r/1000)))
    count += 1000

Average reward per thousand episodes: 

1000 :  0.01900000000000001
2000 :  0.04200000000000003
3000 :  0.2590000000000002
4000 :  0.5020000000000003
5000 :  0.6120000000000004
6000 :  0.6820000000000005
7000 :  0.6620000000000005
8000 :  0.6460000000000005
9000 :  0.6500000000000005
10000 :  0.6640000000000005


In [9]:
# Print updated Q-table
print("\n\n Q-table \n")
print(q_table)



 Q-table 

[[0.5241992  0.48311431 0.43061587 0.44412604]
 [0.19493691 0.         0.02568929 0.08803689]
 [0.25328337 0.04373701 0.         0.06395469]
 [0.         0.         0.         0.        ]
 [0.54049413 0.36068972 0.33956965 0.27675557]
 [0.         0.         0.         0.        ]
 [0.14844535 0.1693424  0.14093882 0.03273721]
 [0.         0.         0.         0.        ]
 [0.47388789 0.40887833 0.4910197  0.56748875]
 [0.36216274 0.59950215 0.46652573 0.2649097 ]
 [0.49475842 0.38083743 0.25956231 0.12847335]
 [0.         0.         0.         0.        ]
 [0.         0.         0.         0.        ]
 [0.25364686 0.53848839 0.73343301 0.43544781]
 [0.73128274 0.86572088 0.78027715 0.68625326]
 [0.         0.         0.         0.        ]]


## Observations:

Over the course of ten thousand episodes, the success rate of the agent increased from 1.9% during its first thousand iterations to 66% during its final thousand.